### ISSUE TITLE
Future provider not updated when using await 

### ISSUE THREAD

**Issue by terryl1900** [rrousselGit/river_pod#1310]
**Describe the bug**
Future provider's value is not updated when using `family.autoDispose` and `await`. In the below example, when clicking the button to update the value, the old value still shows up.

More interestingly, this doesn't work:

```
final dataProvider = StateProvider.family.autoDispose((ref, int id) => 'hello');

final wrapperProvider = FutureProvider.family.autoDispose((ref, int id) async {
  await Future.delayed(const Duration(seconds: 0));
  final data = ref.watch(dataProvider(id));
  return data;
});
```

while these two work properly:

( await after watch)
```
final dataProvider = StateProvider.family.autoDispose((ref, int id) => 'hello');

final wrapperProvider = FutureProvider.family.autoDispose((ref, int id) async {
  final data = ref.watch(dataProvider(id));
  await Future.delayed(const Duration(seconds: 0));
  return data;
});
```

(not using family)
```
final dataProvider = StateProvider((ref) => 'hello');

final wrapperProvider = FutureProvider.family.autoDispose((ref, int id) async {
  await Future.delayed(const Duration(seconds: 0));
  final data = ref.watch(dataProvider);
  return data;
});
```

**To Reproduce**

<details><summary> Full sample code </summary>

```
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

final dataProvider =
    StateProvider.family.autoDispose((ref, int id) => 'hello', name: 'data');

final wrapperProvider = FutureProvider.family.autoDispose((ref, int id) async {
  await Future.delayed(const Duration(seconds: 0));
  final data = ref.watch(dataProvider(id));
  return data.toString();
}, name: 'wrapper');

void main() {
  runApp(const ProviderScope(child: MyApp()));
}

class MyApp extends ConsumerWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final wrappedData = ref.watch(wrapperProvider(42));
    return MaterialApp(
      home: Scaffold(
        body: Center(child: Text(wrappedData.toString())),
        floatingActionButton: FloatingActionButton(
          onPressed: () => ref.read(dataProvider(42).notifier).state = 'bye',
          child: const Icon(Icons.edit),
        ),
      ),
    );
  }
}
```
</details>

I'm using `^2.0.0-dev.4`.

**Comment by rrousselGit** [comment]
The real issue here is likely autoDispose and the fact that you're using watch _after_thr await


It means that during the await, your StateProvider has no listener, and in turn likely got disposed. 

Either increase the disposeDelay on your StateProvider or move the ref.watch before the await

**Comment by terryl1900** [comment]
You are probably right, but I thought the MaterialApp will keep the wrapperProvider provider alive, and wrapperProvider will keep the dataProvider alive, until the widget is disposed of. Should that be the desired behavior?

**Comment by rrousselGit** [comment]
Widgets are unrelated to this.
wrapperProvider is not getting disposed. Only dataProvider is

**Comment by terryl1900** [comment]
Isn't dataProvider being listened by wrapperProvider?

**Comment by rrousselGit** [comment]
Not during the await

The listener is removed when the provider rebuilds. So during the delayed await, it's not listened

**Comment by terryl1900** [comment]
Do you think ideally dependency should be temporarily kept during rebuilds and updated once the rebuild is done?

I can understand this is not easy to do and has low priority. Just want to hear your thought on what's the desired behavior. Feel free to close the issue with an answer.

**Comment by rrousselGit** [comment]
This has been requested before.

The problem is, it'd mean that ref.watch would behave differently based on which provider uses it.

I'm not sure whether that's a good idea, as it could cause significant confusion
